crates/sui-core/src/epoch/reconfiguration.rs

```diff
@@ -38,23 +38,24 @@ where
     pub async fn start_epoch_change(&self) -> SuiResult {
         let epoch = self.state.committee.load().epoch;
         info!(?epoch, "Starting epoch change");
-        if let Some(checkpoints) = &self.state.checkpoints {
-            assert!(
-                checkpoints.lock().is_ready_to_start_epoch_change(),
-                "start_epoch_change called at the wrong checkpoint",
-            );
-        } else {
-            unreachable!();
-        }
+        let checkpoints = self.state.checkpoints.as_ref().unwrap();
+        assert!(
+            checkpoints.lock().is_ready_to_start_epoch_change(),
+            "start_epoch_change called at the wrong checkpoint",
+        );
 
         self.state.halt_validator();
         info!(?epoch, "Validator halted for epoch change");
-        // TODO: The following doesn't work: we also need to make sure that the transactions
-        // all have been included in a batch (and hence will be included in the next checkpoint
-        // proposal).
+        // Check that all transactions that have been sequenced and are about to be committed get
+        // committed. Also make sure that all the transactions that have been committed have made
+        // into a batch. This ensures that they will all made to the next checkpoint proposal.
         // TODO: Use a conditional variable pattern instead of while + sleep.
-        while !self.state.batch_notifier.ticket_drained() {
-            tokio::time::sleep(Duration::from_millis(10)).await;
+        while !self
+            .state
+            .batch_notifier
+            .ticket_drained_til(checkpoints.lock().next_transaction_sequence_expected())
+        {
+            tokio::time::sleep(Duration::from_millis(100)).await;
         }
```

Comment:
This seems dangerous; it it guaranteed to terminate? If so, could we add a comment explaining why (we may forget in a few months)

Reply:
Good question! I assumed this will terminate because every ticket before dropping will increment the low watermark. But I guess things like this is never guaranteed to work 100%. Let me create an issue for this